File: gateway-go/session/session.go
// Package session owns tmux-backed terminal sessions: one pty per session,
// fanned out to every attached channel sink, with the resize-authority and
// scrollback semantics of server/src/terminal-gateway.ts.
//
// Concurrency invariants (the Node gateway got these free from its single
// thread; here they are explicit — spike spec §3):
//   - get-or-create holds the manager mutex: concurrent Attach for a new
//     session spawns exactly one pty.
//   - attached + ring replay + channel subscription happen atomically under
//     the session mutex, so live read-loop output can neither interleave
//     into the replay nor be dropped between replay and subscribe.
//   - Input/Resize are serialized per session by the same mutex.
package session

import (
	"fmt"
	"io"
	"sync"

	"github.com/lean-software-production/ensembleworks/gateway-go/protocol"
)

const (
	minCols, maxCols = 20, 500
	minRows, maxRows = 5, 200
	scrollbackLimit  = 256 * 1024
)

type Pty interface {
	io.ReadWriter
	Resize(cols, rows int) error
	Close() error
}

type PtyFactory func(sessionID string, cols, rows int) (Pty, error)

// Sink is one attached viewer (a relay channel). Implementations must be
// safe to call from the session read-loop goroutine.
type Sink interface {
	SendMsg(inner protocol.Inner)
	SendOutput(payload []byte)
	Close()
}

type sessionState struct {
	mu        sync.Mutex
	id        string
	pty       Pty
	cols, rows int
	ring      [][]byte
	ringBytes int
	channels  map[uint32]Sink
	gone      bool
}

type Manager struct {
	mu       sync.Mutex
	spawn    PtyFactory
	sessions map[string]*sessionState
}

func NewManager(spawn PtyFactory) *Manager {
	return &Manager{spawn: spawn, sessions: make(map[string]*sessionState)}
}

func clamp(v, lo, hi int) int {
	if v < lo {
		return lo
	}
	if v > hi {
		return hi
	}
	return v
}

func (m *Manager) getOrCreate(id string, cols, rows int) (*sessionState, error) {
	m.mu.Lock()
	defer m.mu.Unlock()
	if s, ok := m.sessions[id]; ok {
		return s, nil
	}
	cols, rows = clamp(cols, minCols, maxCols), clamp(rows, minRows, maxRows)
	pty, err := m.spawn(id, cols, rows)
	if err != nil {
		return nil, err
	}
	s := &sessionState{id: id, pty: pty, cols: cols, rows: rows, channels: make(map[uint32]Sink)}
	m.sessions[id] = s
	go m.readLoop(s)
	return s, nil
}

func (m *Manager) readLoop(s *sessionState) {
	buf := make([]byte, 32*1024)
	for {
		n, err := s.pty.Read(buf)
		if n > 0 {
			chunk := append([]byte(nil), buf[:n]...)
			s.mu.Lock()
			s.ring = append(s.ring, chunk)
			s.ringBytes += len(chunk)
			for s.ringBytes > scrollbackLimit && len(s.ring) > 1 {
				s.ringBytes -= len(s.ring[0])
				s.ring = s.ring[1:]
			}
			for _, sink := range s.channels {
				sink.SendOutput(chunk)
			}
			s.mu.Unlock()
		}
		if err != nil {
			s.mu.Lock()
			s.gone = true
			for _, sink := range s.channels {
				sink.SendMsg(protocol.Inner{Type: "exit"})
				sink.Close()
			}
			s.channels = make(map[uint32]Sink)
			s.mu.Unlock()
			m.mu.Lock()
			if m.sessions[s.id] == s {
				delete(m.sessions, s.id)
			}
			m.mu.Unlock()
			return
		}
	}
}

func (m *Manager) Attach(sessionID string, channelID uint32, cols, rows int, sink Sink) error {
	s, err := m.getOrCreate(sessionID, cols, rows)
	if err != nil {
		return err
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.gone {
		return fmt.Errorf("session %s has exited", sessionID)
	}
	// attached carries the SESSION's current size — a newcomer's requested
	// grid must not resize existing viewers (spike spec §2).
	sink.SendMsg(protocol.Inner{Type: "attached", Cols: s.cols, Rows: s.rows})
	for _, chunk := range s.ring {
		sink.SendOutput(chunk)
	}
	s.channels[channelID] = sink
	return nil
}

func (m *Manager) Input(sessionID string, channelID uint32, data string) {
	if s := m.lookup(sessionID); s != nil {
		s.mu.Lock()
		defer s.mu.Unlock()
		if _, attached := s.channels[channelID]; attached && !s.gone {
			_, _ = s.pty.Write([]byte(data))
		}
	}
}

func (m *Manager) Resize(sessionID string, cols, rows int) {
	s := m.lookup(sessionID)
	if s == nil {
		return
	}
	cols, rows = clamp(cols, minCols, maxCols), clamp(rows, minRows, maxRows)
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.gone || (cols == s.cols && rows == s.rows) {
		return // dedup: no pty call, no broadcast (client grid logic relies on this)
	}
	s.cols, s.rows = cols, rows
	_ = s.pty.Resize(cols, rows)
	for _, sink := range s.channels {
		sink.SendMsg(protocol.Inner{Type: "resize", Cols: cols, Rows: rows})
	}
}

func (m *Manager) Detach(sessionID string, channelID uint32) {
	if s := m.lookup(sessionID); s != nil {
		s.mu.Lock()
		delete(s.channels, channelID)
		s.mu.Unlock()
	}
}

// DetachAll drops every viewer (relay disconnect). The ptys stay running —
// tmux sessions must survive connector↔canvas link failures.
func (m *Manager) DetachAll() {
	m.mu.Lock()
	all := make([]*sessionState, 0, len(m.sessions))
	for _, s := range m.sessions {
		all = append(all, s)
	}
	m.mu.Unlock()
	for _, s := range all {
		s.mu.Lock()
		for _, sink := range s.channels {
			sink.Close()
		}
		s.channels = make(map[uint32]Sink)
		s.mu.Unlock()
	}
}

func (m *Manager) lookup(id string) *sessionState {
	m.mu.Lock()
	defer m.mu.Unlock()
	return m.sessions[id]
}

File: cli/src/connector/session.ts
/**
 * ConnectorSessionManager — a faithful TypeScript port of
 * gateway-go/session/session.go's Manager over the shared openTmuxSession
 * primitive. One TmuxSession per canvas session, fanned out to every attached
 * relay channel, with the resize-authority + scrollback-replay + exit-broadcast
 * semantics of session.go (which itself mirrors terminal-gateway.ts).
 *
 * Go's three mutex-protected invariants (get-or-create spawns exactly one pty;
 * attach's attached→replay→subscribe is atomic; input/resize are serialized)
 * hold for free in single-threaded Bun: openTmuxSession is synchronous and the
 * onData read-loop callback plus the mux handlers all run to completion on one
 * event-loop turn, so replay can neither interleave with live output nor drop
 * bytes. The `gone` flag is a defensive mirror (onExit synchronously deletes the
 * session before any later attach), preserved for parity.
 *
 * Trust boundary: the connector deliberately does NOT re-validate sessionId from
 * relay frames — the splicer's ID_RE gate (gateway-registry.ts:184,
 * [a-zA-Z0-9_-]{1,48}) rejects any bad id before a relay-open can exist, and
 * tmux exec-array semantics with the fixed `canvas-` prefix leave no
 * flag-injection surface. This is parity with gateway-go.
 */
import type { TermServerMessage } from '@ensembleworks/contracts'
import { clampTmuxGrid, type TmuxSession } from '@ensembleworks/contracts/session-manager'
import { capTail, type LayoutSnapshot } from './layout.ts'

const SCROLLBACK_LIMIT = 256 * 1024 // bytes replayed to a newly attached channel (session.go)

/** One attached viewer (a relay channel). Down-messages are the inner terminal
 *  protocol (attached/resize/exit); output is raw pty bytes; close tears down. */
export interface ChannelSink {
	sendMsg(inner: TermServerMessage): void
	sendOutput(payload: Buffer): void
	close(): void
}

export type SpawnFactory = (sessionId: string, cols: number, rows: number, cwd?: string) => TmuxSession

interface SessionState {
	pty: TmuxSession
	ring: Buffer[]
	ringBytes: number
	channels: Map<number, ChannelSink>
	gone: boolean
}

export class ConnectorSessionManager {
	private sessions = new Map<string, SessionState>()
	/** SP4 layout seeds, consumed by getOrCreate: last cwd + history to preload
	 *  into the ring. Entries are deleted once used. */
	private seeded = new Map<string, { cwd?: string; history: Buffer }>()
	constructor(private readonly spawn: SpawnFactory) {}

	private getOrCreate(id: string, cols: number, rows: number): SessionState {
		const existing = this.sessions.get(id)
		if (existing) return existing
		const grid = clampTmuxGrid(cols, rows) // session.go getOrCreate clamps BEFORE spawn; attached reports the clamped grid
		const seed = this.seeded.get(id)
		this.seeded.delete(id)
		const pty = this.spawn(id, grid.cols, grid.rows, seed?.cwd) // canvasTmuxSpawnSpec inside; -A reattaches
		const s: SessionState = {
			pty,
			ring: seed && seed.history.byteLength > 0 ? [seed.history] : [],
			ringBytes: seed?.history.byteLength ?? 0,
			channels: new Map(),
			gone: false,
		}
		pty.onData((data) => {
			const buf = Buffer.from(data, 'utf8')
			s.ring.push(buf)
			s.ringBytes += buf.byteLength
			while (s.ringBytes > SCROLLBACK_LIMIT && s.ring.length > 1) s.ringBytes -= s.ring.shift()!.byteLength
			for (const sink of s.channels.values()) sink.sendOutput(buf)
		})
		pty.onExit(() => {
			s.gone = true
			for (const sink of s.channels.values()) {
				sink.sendMsg({ type: 'exit' })
				sink.close()
			}
			s.channels.clear()
			if (this.sessions.get(id) === s) this.sessions.delete(id)
		})
		this.sessions.set(id, s)
		return s
	}

	/** attached carries the SESSION's size (a newcomer's grid must not resize
	 *  existing viewers), then the scrollback ring, then subscribe. */
	attach(id: string, channelId: number, cols: number, rows: number, sink: ChannelSink): boolean {
		const s = this.getOrCreate(id, cols, rows)
		if (s.gone) return false
		sink.sendMsg({ type: 'attached', cols: s.pty.cols, rows: s.pty.rows })
		for (const chunk of s.ring) sink.sendOutput(chunk)
		s.channels.set(channelId, sink)
		return true
	}

	input(id: string, channelId: number, data: string): void {
		const s = this.sessions.get(id)
		if (s && !s.gone && s.channels.has(channelId)) s.pty.write(data)
	}

	/** Session-authoritative resize: TmuxSession.resize clamps+dedups and returns
	 *  whether the grid changed; only then fan out the new size to every viewer. */
	resize(id: string, cols: number, rows: number): void {
		const s = this.sessions.get(id)
		if (!s || s.gone) return
		if (!s.pty.resize(cols, rows)) return
		for (const sink of s.channels.values()) sink.sendMsg({ type: 'resize', cols: s.pty.cols, rows: s.pty.rows })
	}

	detach(id: string, channelId: number): void {
		this.sessions.get(id)?.channels.delete(channelId)
	}

	/** Relay disconnect: drop every viewer, keep the ptys — tmux sessions must
	 *  survive connector↔canvas link failures (session.go DetachAll). On process
	 *  exit the ptys (tmux CLIENTS) get SIGHUP and detach; the tmux server keeps
	 *  the sessions alive for the next connect (-A reattaches). */
	detachAll(): void {
		for (const s of this.sessions.values()) {
			for (const sink of s.channels.values()) sink.close()
			s.channels.clear()
		}
	}

	/** SP4 snapshot (decision #4): every LIVE session's id, cwd (via the
	 *  injected /proc reader — impure, so injected) and capped scrollback tail,
	 *  base64'd for JSON. Called from the connector's SIGTERM handler. */
	snapshotLayout(readCwd: (pid: number | undefined) => string | undefined): LayoutSnapshot {
		const sessions = [...this.sessions.entries()]
			.filter(([, s]) => !s.gone)
			.map(([id, s]) => {
				const cwd = readCwd(s.pty.pid)
				const scrollbackTail = capTail(s.ring).toString('base64')
				return cwd === undefined ? { id, scrollbackTail } : { id, cwd, scrollbackTail }
			})
		return { version: 1, sessions }
	}

	/** SP4 restore (decision #4): record each entry's cwd + history seed, then
	 *  eagerly respawn it at the default 80x24 grid so the session exists (in
	 *  its last cwd, ring pre-seeded) before any viewer attaches. A seed whose
	 *  spawn fails (e.g. its cwd was deleted) is dropped — the next attach
	 *  spawns it fresh with no override; the remaining seeds still restore. */
	preseedLayout(layout: LayoutSnapshot): void {
		for (const entry of layout.sessions) {
			this.seeded.set(entry.id, { cwd: entry.cwd, history: Buffer.from(entry.scrollbackTail, 'base64') })
			try {
				this.getOrCreate(entry.id, 80, 24)
			} catch {
				this.seeded.delete(entry.id) // consumed-or-dropped either way; never fatal
			}
		}
	}
}
